### PR TITLE
Support for playing the alarm on the H100 hub

### PR DESCRIPTION
**Pull request by kay**
See https://github.com/mihai-dinculescu/tapo/issues/310

I used a pattern similar to the `set()` API which I felt was appropriate since you might prefer to use the configured values for some settings.

I've added `stop_alarm` and a method to get a list of available alarm types (ringtones), `get_supported_alarm_type_list`.

I discovered a `Mute` volume option. This isn't available in the tapo app, but using it did make the app bell animation run but no d from the hub. I tried putting through some other random string in the `alarm_volume` parameter, and it used the configured volume, so apparently `Mute` is a valid value but I am not sure what use it is. I've included this in the PR for completeness but can remove it if preferred.

### CODE REVIEW

**Comment by mihai-dinculescu** [comment]
This is a very solid contribution. Thank you! I particularly appreciate the clarity of the description and the cleanliness of the implementation.

I have three comments:

**Mute**: I tend to agree with you. If we don't understand its usefulness and no one is asking for it, it's best left out. Considering that it's not used by the Tapo app, my guess is that it's a feature that was left half-implemented. It might be useful for silent alarms, but I couldn't figure out how.

**Ergonomics**: I think that the builder pattern is too much in this case, and there are ways in which the API could be simplified. One possible simplification might be to have a `play_alarm` method that always takes three parameters: `type`, `volume`, and `duration`. To maintain flexibility, `volume` could get a fourth option, `Default`, and `type` could become an enum that contains the values returned by `get_supported_alarm_type_list` + `Default`.

This would simplify:

```rust
    hub.play_alarm()
        .with_alarm_type("Hardcoded string alarm type")
        .with_alarm_volume(AlarmVolume::Low)
        .with_alarm_duration(1)
        .send(&hub)
        .await?;
 ```
to:

rust
```
hub.play_alarm(AlarmType::Default, AlarmVolume::Low, 1).await?;
```
and it would eliminate the need to call `get_supported_alarm_type_list` while maintaining the same flexibility by fully utilizing Rust's type system.

**Durations**: The standard in Rust is to take `duration: std::time::Duration` as a parameter rather than numeric values for seconds, minutes, etc.

What do you think?

**Comment by kay** [comment]
On `Mute`, will remove.

The reason I didn't use `Duration` was that the granularity is limited to seconds. So if you have `Duration::from_millis(999)` you would want it either to round up or down, but `as_secs` gives whole seconds only. In my testing, a duration 0 behaved differently than not passing in any duration. For some alarm types it seemed to stop after one play of the audio track, maybe 5-6 seconds, but others seem to play forever (or at least as long as my partner will put up with me making noise). 

I did think I saw some stuff in the decompilation that maybe it plays by default for 5 minutes. I'd have to do further testing to find it's exact behaviour, so that I can handle non-zero, but sub-second `Duration` inputs. So I side stepped the problem by allowing it to be omitted. Maybe another option is to forbid non-zero, sub second durations with an `Err`? What do you think? 

On eliminating `get_supported_alarm_type_list `, I wasn't sure what markets Tapo sells to, I wasn't sure if the strings it returns are fixed, subject to internationalisation, or subject to change in future firmware updates. I could imagine some logic being, "I want to play 'Alarm 1' but if it is not available, pick any valid ringtone". This is useful as home alarm to ensure that _some_ sound plays. If `AlarmType::Default` was introduced this would help by deferring to the hub configuration but isn't quite the same as it means you can't have two separate types of automation using the same hub but triggering different sounds, with the same fallback. 

On explicitly defining all the ringtones as variants, I didn't want to create a maintenance headache for you or users if anything changed. Potentially it could result in users not being able to work around the issue until a new version of this crate is released, and what happens with users who want to keep the old firmware working? Maybe we can have `AlarmType::Custom` that accepts any string, and I can predefine all the variants I've discovered so far, but how do users discover other valid strings if it does change?

On the builder pattern, I think some of these parameters should be optional but I'm not opinionated about how. The alarm type we've talked about but the duration I think should be optional because it avoids the 0 edge case. If people explicitly want it to play forever, they don't need to set it. Not passing any value in the request seems to play forever for all the ringtones I tried, which wasn't them all.

`AlarmVolume` seems fixed and no way to discover the valid strings here that I can find, so I think this one is ok as an enum. I could add a `AlarmVolume::Default` variant to this too to avoid an `Option` here.

So my preference given your feedback is for a plain `play_alarm` function, returning a `Result<(), Error>` (not a builder) where
1. `duration` is either
  a. `Option<Duration>`
  b.  `Duration` where can return `Error::Validation` if duration > 0 and < 1sec (avoids user surprise)
  c.  `Duration` where can return `Error::Validation` if duration is not whole seconds (avoids user surprise)
  d.  `Duration` where can return `Error::Validation` if duration < 1sec (since I can't document what 0 secs actually does yet)
2. `alarm_type` is either
  a. `Option<String>`
  b. `AlarmType` where `Custom` and `Default` are defined as the only variants (low maintenance on your part/this crate maintainers)
  c. `AlarmType` where `Custom` and `Default` are defined as variants, and all others I've discovered so far are too (this lets you use the type system but offers the option of graceful fallback)
3.  `volume` is `AlarmVolume`, but without `Mute` as  an option

I look forward to hearing your thoughts

**Comment by mihai-dinculescu** [comment]
> On Mute, will remove.

I think I misunderstood your initial point about `Mute`. If it makes the bell pop in the app and likely makes `get_device_info` return `in_alarm: true`, it’s worth keeping as a functional silent alarm.

> The reason I didn't use Duration was that the granularity is limited to seconds. So if you have Duration::from_millis(999) you would want it either to round up or down, but as_secs gives whole seconds only. In my testing, a duration 0 behaved differently than not passing in any duration. For some alarm types it seemed to stop after one play of the audio track, maybe 5-6 seconds, but others seem to play forever (or at least as long as my partner will put up with me making noise).
> 
> I did think I saw some stuff in the decompilation that maybe it plays by default for 5 minutes. I'd have to do further testing to find it's exact behaviour, so that I can handle non-zero, but sub-second Duration inputs. So I side stepped the problem by allowing it to be omitted. Maybe another option is to forbid non-zero, sub second durations with an Err? What do you think?

I see what you mean. Using `Duration` is quite awkward in this case. I'm happy to use a number of seconds instead.
Which option causes some ringtones to play indefinitely? Is it None or 0? That might be worth supporting.

> On eliminating get_supported_alarm_type_list , I wasn't sure what markets Tapo sells to, I wasn't sure if the strings it returns are fixed, subject to internationalisation, or subject to change in future firmware updates. I could imagine some logic being, "I want to play 'Alarm 1' but if it is not available, pick any valid ringtone". This is useful as home alarm to ensure that some sound plays. If AlarmType::Default was introduced this would help by deferring to the hub configuration but isn't quite the same as it means you can't have two separate types of automation using the same hub but triggering different sounds, with the same fallback.

`get_supported_alarm_type_list` is definitely worth keeping in the library as a debug/information-gathering method. Ideally, it should be behind a `debug` feature flag along with a few other methods, but this is for another PR. I wasn't suggesting its removal, but rather not forcing the user to use it to figure out the available options (which can be better provided by an enum).

> On explicitly defining all the ringtones as variants, I didn't want to create a maintenance headache for you or users if anything changed. Potentially it could result in users not being able to work around the issue until a new version of this crate is released, and what happens with users who want to keep the old firmware working? Maybe we can have AlarmType::Custom that accepts any string, and I can predefine all the variants I've discovered so far, but how do users discover other valid strings if it does change?

The design philosophy so far has been to take the time to understand how the Tapo API works and leverage the extraordinarily powerful Rust type system to provide a highly ergonomic and self-explanatory API, making it easy for the user to choose between only valid options. Otherwise, the choice of Rust wouldn't make sense, and it would just be another library that forces the user to understand the inner workings of the Tapo API in depth.

This comes with the trade-off you mentioned, but I believe it's worth it if it provides users with a very pleasant and easy-to-use interface.

Regarding versioning, my intention is to keep the latest version of the library compatible with the latest firmware version. If, for any reason, a user chooses not to update their devices, they can continue using the previous version that was compatible with it. (Note to self to include something along these lines in the CONTRIBUTING.md)

I might have been overthinking it with `Default`. It's cute, but it requires the user to import the enum for nothing. `Option` might be a simpler and more obvious choice.

WDYT of `async fn play_alarm(seconds: u32, ringtone: Option<Ringtone>, volume:  Option<AlarmVolume>) -> Result<(), Error>`, pending the open question about seconds?

**Comment by kay** [comment]
> Which option causes some ringtones to play indefinitely? Is it None or 0? That might be worth supporting.

`None` definitely seems to allow all the ringtones I played with to play forever.

> it should be behind a `debug` feature flag along with a few other methods, but this is for another PR.

Happy to compile this out if you like, just let me know if you wanted a feature flag, or to use the build profile `debug_assertions` setting, or if you want me to leave it alone for now?

> WDYT of async fn play_alarm(seconds: u32, ringtone: Option<Ringtone>, volume:  Option<AlarmVolume>) -> Result<(), Error>, pending the open question about seconds?

I'd suggest allowing seconds to be optional because its absence would suite my use case more, either that or I'd have to see what the max value I can pass for duration. So I'd suggest `async fn play_alarm(seconds: Option<u32>, ringtone: Option<Ringtone>, volume:  Option<AlarmVolume>) -> Result<(), Error>`.

> Regarding versioning, my intention is to keep the latest version of the library compatible with the latest firmware version

This is fine if the user of this crate keeps up with this crate & the tapo firmware, but it'd suck to have a firmware update be automatically picked up by the devices, and the alarm gets triggered by the user software, using an old version of this crate, and finds out that tapo doesn't support that alarm type anymore and nothing happens. 
It is very edge case, I can't imagine tapo changing their supported alarms, and I'll accept your decision here, but to clarify, on `Ringtone`, are you comfortable if I add an `Custom` variant _in addition_ to all the known strings? So that users _can_ write robust code that has a fallback if they want to.

**Comment by kay** [comment]
Would possibly an `impl Into<Ringtone>` work as a parameter? With a `impl From<String> for Ringtone` perhaps?

**Comment by kay** [comment]
I checked what happens if I send an unexpected string in and it comes back with `Error: Tapo(Unknown(-1008))` this is probably good enough actually to cover my use case. The solution would be, try with the custom tone, and if you get an error, try with the configured tone. So I don't need a `Custom` or anything like that.

**Comment by kay** [comment]
> I think I misunderstood your initial point about `Mute`. If it makes the bell pop in the app and likely makes `get_device_info` return `in_alarm: true`, it’s worth keeping as a functional silent alarm.

I checked this, and you're right. Without the alarm you get `in_alarm_source: "", in_alarm: false`. WIth the alarm going off you get `in_alarm_source: "", in_alarm: false` so I'll leave it in.

**Comment by mihai-dinculescu** [comment]
> Happy to compile this out if you like, just let me know if you wanted a feature flag, or to use the build profile debug_assertions setting, or if you want me to leave it alone for now?

I believe that get_supported_alarm_type_list can remain as is, and we can discuss the separation in a separate issue or PR. This matter is distinct and will require some design and consideration regarding how it should function in Python. Of course, you're more than welcome to contribute to this effort.

> I'd suggest allowing seconds to be optional because its absence would suite my use case more, either that or I'd have to see what the max value I can pass for duration. So I'd suggest `async fn play_alarm(seconds: Option<u32>, ringtone: Option<Ringtone>, volume:  Option<AlarmVolume>) -> Result<(), Error>`.

👍 With the comment that if the duration is optional, it should probably come last. Also, `AlarmRingtone` might be a better name for the enum as it's consistent with `AlarmVolume`.

> It is very edge case, I can't imagine tapo changing their supported alarms, and I'll accept your decision here, but to clarify, on Ringtone, are you comfortable if I add an Custom variant in addition to all the known strings? So that users can write robust code that has a fallback if they want to.

I'm hesitant to add `Custom` until we receive more feedback, as it seems to be a lesser version of the enum options, and the duplication might create confusion. If a custom string is provided that doesn't exist on the device or stops existing on the device, it will result in an error (LE you've just pointed one the same thing). Although handling the error by changing the custom string will be easier than waiting for a new library version or forking it, it still presents an inconvenience.

If feedback comes in that some H100 devices have a different list of alarms for any reason or if TP-Link frequently changes the alarm options, that would be a compelling reason to add it.

In a nutshell, it might be a premature optimization.

> Would possibly an impl Into<Ringtone> work as a parameter? With a impl From<String> for Ringtone perhaps?
This would work if enums supported private variants, but unfortunately, they didn't.

**Comment by kay** [comment]
I wrote some code to run and use the device info to measure how long each alarm runs when you give an alarm_duration=0. I discovered that `in_alarm` appears to go back to being `false` almost immediately, but the sound continues to play until the end of the track. I feel like 0 might be a real edge case with this. If I put a non-zero duration longer or shorter than the length of the audio track, and again use `in_alarm` it always plays for a least the requested duration. 

Examples are `Doorbell Ring 1` plays for approx 4-5 seconds when duration = 0. Phone Ring about 4 seconds. Alarm 1 about 6 seconds. Alarm 2 about 10 seconds. Alarm 3 about 16 seconds.

**Comment by mihai-dinculescu** [comment]
That's interesting. In that case, would it make sense to contemplate something like `duration: AlarmDuration` where

```rust
enum Duration {
    Continuous,  // none
    Once, // 0
    Seconds(u32),
}
```

**Comment by kay** [comment]
I've pushed my changes based on your feedback. Let me know if you think I should change anything further. I appreciate all your feedback. 

On the duration=0, this smells like a bug in tapo software, I don't know if it is intentional to play the full audio track because of the `in_alarm` flag getting reset after the requested duration, which I think also controls the indication in the app. For the `Mute` volume `Once` wont mean much since it unsets so fast you can't observe it. But I'm happy to adopt your `AlarmDuration` suggestion if you like, just confirm that's what you want and I'll make the change.

I'm also happy to go with your `AlarmDuration` enum without the `Once` variant, but earlier you seemed concerned about unneeded imports, since `Option` is in the prelude you don't need to update it.

**Comment by mihai-dinculescu** [comment]
> I've pushed my changes based on your feedback. Let me know if you think I should change anything further. I appreciate all your feedback.
> 
> On the duration=0, this smells like a bug in tapo software, I don't know if it is intentional to play the full audio track because of the `in_alarm` flag getting reset after the requested duration, which I think also controls the indication in the app. For the `Mute` volume `Once` wont mean much since it unsets so fast you can't observe it. But I'm happy to adopt your `AlarmDuration` suggestion if you like, just confirm that's what you want and I'll make the change.
> 
> I'm also happy to go with your `AlarmDuration` enum without the `Once` variant, but earlier you seemed concerned about unneeded imports, since `Option` is in the prelude you don't need to import it.

The preview in the Tapo App may use 0 to play the tune once.
I'm happy to follow your preference on this. The enum makes sense if there are three values. If there are only two, `Option` is indeed better.

**Comment by kay** [comment]
Good point regarding the preview, I suspect you're right. I tested them from the app and they play for the same rough runtime as I observed earlier.

I don't feel strongly, I did wince a bit at `Option<u32>`. I did consider adding a type alias `type Seconds = u32` so that I could write `Option<u32>`. I like your enum too, but I'm not sure about the `Once`.

If I added your enum, should I allow `Seconds(0)` still?

**Comment by kay** [comment]
I've added the `AlarmDuration` enum. Let me know if you think of anything else.

**Comment by mihai-dinculescu** [comment]
> If I added your enum, should I allow `Seconds(0)` still?

I believe there is a strong case for a validation error, as you previously suggested.

**Comment by kay** [comment]
I've added commits for stronger error handling

File: tapo/examples/tapo_h100.rs [hunk 1]
```diff
@@ -113,5 +115,27 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         }
     }
 
+    let supported_ringtones = hub.get_supported_alarm_type_list().await?;
+    info!("Supported ringtones: {:?}", supported_ringtones);
```
Comment (mihai-dinculescu) [inline tapo/examples/tapo_h100.rs:119]: Is this necessary in the example, given that it's an edge and/or advanced use case?

Comment (kay) [reply tapo/examples/tapo_h100.rs]: Removed

File: tapo/src/requests/play_alarm.rs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    fn validate(&self) -> Result<(), Error> {
+        if let AlarmDuration::Seconds(seconds) = self.alarm_duration {
+            if seconds == 0 {
+                return Err(Error::Validation {
+                    field: "alarm_duration".to_string(),
+                    message: "seconds must be greater than zero".to_string(),
+                });
+            }
+        }
+        Ok(())
```
[the remark this answers is not in the document]
Comment (kay) [reply tapo/src/requests/play_alarm.rs]: Used your snippet, thanks

File: tapo/src/requests/play_alarm.rs [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+/// The volume of the alarm.
+/// For the H100, this is a fixed list of volume levels.
+#[derive(Debug, Default, Serialize)]
+#[cfg_attr(test, derive(Clone, Copy))]
+#[serde(rename_all = "lowercase")]
+pub enum AlarmVolume {
+    /// Mute the audio output from the alarm.
+    /// This causes the alarm in the app to go off but nothing audible from the hub.
+    /// You can check if the alarm is going off from the `in_alarm` field of
+    /// [`crate::DeviceInfoHubResult`].
```
[the remark this answers is not in the document]
Comment (kay) [reply tapo/src/requests/play_alarm.rs]: Thanks, your phrasing is well better

File: tapo/src/api/hub_handler.rs [hunk 4]
```diff
@@ -58,6 +58,37 @@ impl HubHandler {
...
+            .read()
+            .await
+            .get_supported_alarm_type_list()
+            .await
+            .map(|response| response.alarm_type_list)
+    }
+
+    /// Start playing the hub alarm.
+    /// By default, this uses the configured alarm settings on the hub.
+    /// You can override each of the settings by passing `Some` as a parameter.
```
[the remark this answers is not in the document]
Comment (kay) [reply tapo/src/api/hub_handler.rs]: Agreed, fixed

File: tapo/src/api/hub_handler.rs [hunk 5]
```diff
@@ -58,6 +58,37 @@ impl HubHandler {
         self.client.read().await.get_device_info().await
     }
 
+    /// Returns a list of alarm types (ringtones) supported by the hub.
+    pub async fn get_supported_alarm_type_list(&self) -> Result<Vec<String>, Error> {
```
[the remark this answers is not in the document]
Comment (kay) [reply tapo/src/api/hub_handler.rs]: I started to using ringtone based on our conversations, I don't feel strongly either way. Ringtone is what they call it in the app. But "alarm type" is what their API calls it. I've been over (I think) all uses and fixed it such that ringtone is used on any public methods from this crate, but I've left the "alarm_type" name internally since that is what their API calls it. Are you happy with that split?

File: tapo/src/responses/get_support_alarm_type_list_result.rs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+use serde::Deserialize;
+
+use super::TapoResponseExt;
+
+/// Contains a list of supported alarm types (ringtones) of the device.
+#[derive(Debug, Deserialize)]
+pub struct GetSupportAlarmTypeListResult {
+    /// Available alarm types that can be passed to [PlayAlarmParams](`crate::requests::PlayAlarmParams`)
```
[the remark this answers is not in the document]
Comment (kay) [reply tapo/src/responses/get_support_alarm_type_list_result.rs]: I've removed the link to `PlayAlarmParams` params struct, but as this is intended to stick around for debugging, I've instead referred to the name of the Tapo API endpoint instead.

File: tapo/src/responses/get_support_alarm_type_list_result.rs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+use serde::Deserialize;
+
+use super::TapoResponseExt;
+
+/// Contains a list of supported alarm types (ringtones) of the device.
+#[derive(Debug, Deserialize)]
+pub struct GetSupportAlarmTypeListResult {
```
[the remark this answers is not in the document]
Comment (kay) [reply tapo/src/responses/get_support_alarm_type_list_result.rs]: I don't know why I prefixed this. Fix.

File: README.md [hunk 8]
```diff
@@ -49,6 +49,10 @@ Unofficial Tapo API Client. Works with TP-Link Tapo smart devices. Tested with l
...
 | set() API \*                         |                              |                        |          &#x2705;           |      &#x2705;       |        &#x2705;         |                         |                         |                         |                     |
+| get_supported_alarm_type_list        |                              |                        |          &#x2705;           |      &#x2705;       |        &#x2705;         |                         |                         |                         |       &check;       |
+| play_alarm                           |                              |                        |          &#x2705;           |      &#x2705;       |        &#x2705;         |                         |                         |                         |       &check;       |
+| stop_alarm                           |                              |                        |          &#x2705;           |      &#x2705;       |        &#x2705;         |                         |                         |                         |       &check;       |
+
```
[the remark this answers is not in the document]
Comment (kay) [reply README.md]: Fixed. I also spotted that I had `&#x2705;` in 3 columns I shouldn't have had. I've only tested this for H100 and I've no reason to believe it would work for any other device. So I've removed the incorrect checkboxes.

**Comment by kay** [comment]
I think I've addressed all your feedback. I'll leave the conversations open for you to resolve once happy.

**Comment by kay** [comment]
I'd prefer not to work on the Python wrapper. Maybe someone who knows about the toolchains used by the wrapper, or has the time to learn, might help out and finish what I haven't. I did have a look and I wouldn't be able to make good progress with the available free time I have.

**Review by mihai-dinculescu — APPROVED** [review]
